src/confidential.rs

```diff
@@ -17,251 +17,231 @@
-            fn deserialize<D: Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
-                use serde::de::{Error, Visitor, SeqAccess};
-                struct CommitVisitor;
-
-                impl <'de> Visitor<'de> for CommitVisitor {
-                    type Value = $name;
+impl Decodable for Value {
+    fn consensus_decode<D: io::BufRead>(mut d: D) -> Result<Value, encode::Error> {
+        let prefix = {
+            let buffer = d.fill_buf()?;
 
-                    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
-                        f.write_str("a committed value")
-                    }
+            if buffer.is_empty() {
+                return Err(encode::Error::UnexpectedEOF);
+            }
 
-                    fn visit_seq<A: SeqAccess<'de>>(self, mut access: A) -> Result<Self::Value, A::Error> {
-                        let prefix: u8 = if let Some(x) = access.next_element()? {
-                            x
-                        } else {
-                            return Err(A::Error::custom("missing prefix"));
-                        };
-
-                        match prefix {
-                            0 => Ok($name::Null),
-                            1 => {
-                                // Apply $explicit_fn to allow `Value` to swap the amount bytes
-                                match access.next_element()? {
-                                    Some(x) => Ok($name::Explicit($explicit_fn(x))),
-                                    None => Err(A::Error::custom("missing commitment")),
-                                }
-                            }
-                            p if p == $prefixA || p == $prefixB => {
-                                match access.next_element()? {
-                                    Some(y) => Ok($name::Confidential(p, y)),
-                                    None => Err(A::Error::custom("missing commitment")),
-                                }
-                            }
-                            p => return Err(A::Error::custom(format!(
-                                "invalid commitment, invalid prefix: 0x{:02x}", p
-                            ))),
-                        }
-                    }
-                }
+            buffer[0]
+        };
 
-                d.deserialize_seq(CommitVisitor)
+        match prefix {
+            0 => {
+                // consume null value prefix
+                d.consume(1);
+                Ok(Value::Null)
+            }
+            1 => {
+                // ignore prefix when decoding an explicit value
+                d.consume(1);
+                let explicit = u64::swap_bytes(Decodable::consensus_decode(&mut d)?);
+                Ok(Value::Explicit(explicit))
+            }
+            p if p == 0x08 || p == 0x09 => {
+                let bytes = <[u8; 33]>::consensus_decode(&mut d)?;
+                Ok(Value::Confidential(PedersenCommitment::from_slice(&bytes)?))
```

Comment:
You could also implement `Decodable` for `PedersenCommitment` and then simply delegate here. With `BufRead`, we can still peek at the prefix and use it to decide on what to delegate to while retaining the prefix in the buffer so the implementation can use it to decode.